graph_coloring.py

```
@@ -12,78 +12,107 @@
-nx.draw(G, pos=pos, node_color=node_colors, node_size=50, edgecolors='k', cmap='hsv')
-plt.savefig('graph_result.png')
-
-# check that colors are different
-valid = True
-for edge in G.edges:
-    i, j = edge
-    if sample[i] == sample[j]:
-        valid = False
-        break
-print("\nSolution validity: ", valid)
-
-colors_used = max(sample.values())+1
-print("\nColors required:", colors_used, "\n")
+def build_graph(num_nodes):
+    """Build graph."""
+
+    print("\nBuilding graph...")
+
+    G = nx.powerlaw_cluster_graph(num_nodes, 3, 0.4)
+    pos = nx.spring_layout(G)
+    nx.draw(G, pos=pos, node_size=50, edgecolors='k', cmap='hsv')
+    plt.savefig("original_graph.png")
+    plt.clf()
+
+    return G, pos
+
+def build_cqm(G, num_colors):
+    """Build CQM model."""
+
+    print("\nBuilding constrained quadratic model...")
+
+    # Initialize the CQM object
+    cqm = ConstrainedQuadraticModel()
+
+    # Build CQM variables
+    colors = {n: {c: Binary((n,c)) for c in range(num_colors)} for n in G.nodes}
+
+    # Build the objective: sum all colors values
+    cqm.set_objective(quicksum([c*colors[n][c] for n in G.nodes for c in range(num_colors)]))
+
+    # Add constraint to make variables discrete
+    for n in G.nodes():
+        cqm.add_discrete([(n,c) for c in range(num_colors)])
+  
+    # Build the constraints: edges have different color end points
+    for u, v in G.edges:
+        for c in range(num_colors):
+            cqm.add_constraint(colors[u][c]*colors[v][c] == 0)
+
+    return cqm
+
+def run_hybrid_solver(cqm):
+    """Solve CQM using hybrid solver."""
+
+    print("\nRunning hybrid sampler...")
+
+    # Initialize the CQM solver
+    sampler = LeapHybridCQMSampler()
+
+    # Solve the problem using the CQM solver
+    sampleset = sampler.sample_cqm(cqm, label='Example - Graph Coloring')
+    feasible_sampleset = sampleset.filter(lambda row: row.is_feasible)
+
+    ss = feasible_sampleset.first.sample
```

What about nicer error handling for the (unlikely) case the solver fails to find a feasible solution rather than leaving it to break and return a Python error message?

Also, "ss" is not the best name in public example